internal/langserver/handlers/command/module_list.go

```diff
@@ -0,0 +1,145 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/creachadair/jrpc2/code"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	"github.com/hashicorp/terraform-ls/internal/langserver/cmd"
+	"github.com/hashicorp/terraform-ls/internal/terraform/datadir"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+const moduleListVersion = 0
+
+type moduleListResponse struct {
+	FormatVersion int               `json:"v"`
+	Modules       []terraformModule `json:"modules"`
+}
+
+type terraformModule struct {
+	Name             string            `json:"name"`
+	Path             string            `json:"path"`
+	Version          string            `json:"version"`
+	Type             string            `json:"type"`
+	DocLink          string            `json:"docklink"`
+	DependentModules []terraformModule `json:"depmodules"`
+}
+
+func ModuleListHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return nil, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return nil, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return nil, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found.ModManifest == nil {
+		return nil, nil
+	}
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
+			// this is the current directory, which is technically a module
+			// skipping as it's not relevant in the activity bar (yet?)
+			continue
+		}
+
+		moduleName := manifest.Key
+		subModuleName := ""
+
+		// determine if this module is nested in another module
+		// in the currecnt workspace by finding a period in the moduleName
+		// is it better to look at SourceAddr and compare?
+		if strings.Contains(manifest.Key, ".") {
+			v := strings.Split(manifest.Key, ".")
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		// github modules are put in `files` directory, but Manifest Record does not
+		// record the `name` in Key it puts it in SourceAddr
+		if strings.HasPrefix(manifest.Key, "files") {
+			moduleName = manifest.SourceAddr
+		}
+
+		// build what we know
+		moduleInfo := terraformModule{
+			Name:    moduleName,
+			Path:    strings.TrimSuffix(manifest.SourceAddr, "/"),
```

Comment:
I think we're mixing two different things here.

As mentioned, `manifest.SourceAddr` is just the "human-typed" address, e.g. `github.com/hashicorp/example` or `hashicorp/consul/aws` (Registry address), it's not strictly speaking a path, but it's probably useful to display in the UI. I guess we can pass this as a separate field `SourceAddr` here?

`manifest.Path` is where the module was actually installed on the filesystem. That is probably _not_ something you would display in the UI, but it may be useful if we want to give users the option to "open installed module" for any local modules? Maybe we can omit `Path` initially.

Reply:
I renamed `path` to `source` to more accurately reflect the idea I was going for in the VS Code UI. Going to omit `manifest.Path` for now, as a TODO for later when we get some user feedback